File: information-security/port-scanner/port_scanner.py
import re
import socket
from typing import List, Union

from common_ports import ports_and_services

ERROR_INVALID_HOSTNAME = "Error: Invalid hostname"
ERROR_INVALID_IP_ADDRESS = "Error: Invalid IP address"


def get_open_ports(target: str, port_range: List[int], is_verbose: bool = False) -> Union[list, str]:
    if is_ip(target):
        host, address = "", target
        try:
            socket.inet_pton(socket.AF_INET, address)
        except socket.error:
            return ERROR_INVALID_IP_ADDRESS

        try:
            host = socket.gethostbyaddr(address)
            if type(host) is tuple:
                host = host[0]
        except socket.error:
            host = ""
    else:
        host, address = target, ""
        try:
            address = socket.gethostbyname(host)
        except socket.gaierror:
            return ERROR_INVALID_HOSTNAME

    open_ports = []
    start, stop = port_range
    for port in range(start, stop+1):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.settimeout(1)
        try:
            s.connect((target, port))
            open_ports.append(port)
        except socket.error:
            pass
        finally:
            s.close()

    if not is_verbose:
        return open_ports

    res = f"Open ports for "
    if host:
        res += f"{host}"
    if address and not host:
        res += f"{address}"
    elif address:
        res += f" ({address})"
    res += "\nPORT     SERVICE\n"
    res += "\n".join(map(lambda p: "{:<9d}{}".format(p, ports_and_services.get(p)), open_ports))
    return res


def is_ip(target: str) -> bool:
    m = re.match(r"^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$", target)
    return bool(m) and all

File: information-security/port-scanner/portscanner.go
package main

import (
	"context"
	"errors"
	"fmt"
	"net"
	"net/netip"
	"strconv"
	"strings"
	"time"
)

const (
	scanOpenPortsTimeout = 60 * time.Second

	lookupTimeout = time.Second

	scanPortTimeout = 500 * time.Millisecond
	scanPortNetwork = "tcp"
)

var (
	ErrorInvalidHostname  = errors.New("invalid hostname")
	ErrorInvalidIPAddress = errors.New("invalid IP address")
)

type OpenPortsResult struct {
	ports   []int
	host    string
	address string
}

func OpenPorts(target string, portStart, portEnd int) (OpenPortsResult, error) {
	if target == "" {
		return OpenPortsResult{}, errors.New("target must be non-empty")
	}
	if portStart < 0 || portEnd < 0 {
		return OpenPortsResult{}, errors.New("ports must be positive")
	}
	if portStart > portEnd {
		return OpenPortsResult{}, errors.New("portStart must be lower portEnd")
	}

	ctx, cancel := context.WithTimeout(context.Background(), scanOpenPortsTimeout)
	defer cancel()

	var host, address string
	if ifCanBeIP4(target) {
		address = target
		h, err := lookupHost(ctx, address)
		if err != nil {
			return OpenPortsResult{}, err
		}
		host = h
	} else {
		host = target
		addr, err := lookupAddress(ctx, host)
		if err != nil {
			return OpenPortsResult{}, err
		}
		address = addr
	}

	ports := make([]int, 0, portEnd-portStart)
	for port := portStart; port <= portEnd; port++ {
		if scanPort(ctx, target, port) {
			ports = append(ports, port)
		}
	}

	if ctx.Err() != nil {
		return OpenPortsResult{}, ctx.Err()
	}

	return OpenPortsResult{
		ports:   ports,
		host:    host,
		address: address,
	}, nil
}

func (o *OpenPortsResult) Ports() []int {
	return o.ports
}

func (o *OpenPortsResult) Verbose() string {
	var sb strings.Builder
	sb.Grow(100)
	_, _ = sb.WriteString("Open ports for ")
	if o.host != "" {
		_, _ = sb.WriteString(o.host)
	}
	if o.address != "" && o.host == "" {
		_, _ = sb.WriteString(o.address)
	} else if o.address != "" {
		_, _ = sb.WriteString(" (")
		_, _ = sb.WriteString(o.address)
		_, _ = sb.WriteString(")")
	}
	_, _ = sb.WriteString("\nPORT     SERVICE")
	for _, p := range o.ports {
		port := fmt.Sprintf("\n%-9d", p)
		_, _ = sb.WriteString(port)
		_, _ = sb.WriteString(portsServices[p])
	}

	return sb.String()
}

func lookupHost(ctx context.Context, address string) (string, error) {
	if _, err := netip.ParseAddr(address); err != nil {
		return "", ErrorInvalidIPAddress
	}

	lookupCtx, lookupCancel := context.WithTimeout(ctx, lookupTimeout)
	defer lookupCancel()

	names, err := net.DefaultResolver.LookupAddr(lookupCtx, address)
	if err == nil && len(names) > 0 {
		return strings.TrimSuffix(names[0], "."), nil
	}

	return "", nil
}

func lookupAddress(ctx context.Context, host string) (string, error) {
	lookupCtx, lookupCancel := context.WithTimeout(ctx, lookupTimeout)
	defer lookupCancel()

	addresses, err := net.DefaultResolver.LookupHost(lookupCtx, host)
	if err != nil {
		return "", ErrorInvalidHostname
	}

	if len(addresses) > 0 {
		return addresses[0], nil
	}

	return "", nil
}

func scanPort(ctx context.Context, target string, port int) bool {
	scanCtx, scanCancel := context.WithTimeout(ctx, scanPortTimeout)
	defer scanCancel()

	address := target + ":" + strconv.Itoa(port)

	var d net.Dialer
	conn, err := d.DialContext(scanCtx, scanPortNetwork, address)
	if err != nil {
		return false
	}
	defer func() { _ = conn.Close() }()

	return true
}

func ifCanBeIP4(target string) bool {
	octets := strings.Split(target, ".")
	if len(octets) != 4 {
		return false
	}
	for _, oct := range octets {
		_, err := strconv.Atoi(oct)
		if err != nil {
			return false
		}
	}
	return true
}
